### ISSUE TITLE
Failed to submit array

### ISSUE THREAD

**Issue by bbarbazuk** [marbl/canu#801]
```
I am using CANU 1.6 manually doing each of the three steps. I managed to get the correction step done, and its now part way through trimming.  It has crashed though – and it seems to be an issue with QOS. 
Going through the logs, this is the main crash report detail:
 
The error thrown is:
CRASH: Canu 1.6
CRASH: Please panic, this is abnormal.
ABORT:
CRASH:   Failed to submit batch jobs.
CRASH:
CRASH: Failed at /apps/canu/1.6/bin/lib/canu/Execution.pm line 1119.
CRASH: canu::Execution::submitOrRunParallelJob('Spiny_02022018', 'obtovl', 'trimming/1-overlapper', 'overlap', 1, 2, 3, 4, 5, ...)
called at /apps/canu/1.6/bin/lib/canu/OverlapInCore.pm line 429
CRASH: canu::OverlapInCore::overlapCheck('Spiny_02022018', 'obt', 'partial') called at /apps/canu/1.6/bin/canu line 638
CRASH: main::overlap('Spiny_02022018', 'obt') called at /apps/canu/1.6/bin/canu line 698
CRASH:
CRASH: Last 50 lines of the relevant log file (trimming/1-overlapper/overlap.jobSubmit-01.out):
CRASH:
 
I had a look at the script it was tring to run when the CRASH occurred as well as the associated log.
 
The script: trimming/1-overlapper/overlap.jobSubmit-01.sh requests (note that canu generated this:
#!/bin/sh
 
sbatch \
  --mem-per-cpu=1g --cpus-per-task=4 --mem=128gb --time=12-00:00:00 --qos=barbazuk -o overlap.%A_%a.out \
  -D `pwd` -J "obtovl_Spiny_02022018_CANU-default" \
  -a 1-3000 \
  ./overlap.sh 0 \
> ./overlap.jobSubmit-01.out 2>&1
 
 
and the trimming/1-overlapper/overlap.jobSubmit-01.out reports:
sbatch: error: Batch job submission failed: Job violates accounting/QOS policy (job submit limit, user's size and/or time limits).
 
To my untrained eye this doesn’t look horrible.  The burst queue has a 4 day limit, by I have access to over 3000 cores and 13Gb aggregated RAM assuming its available.  The array job limit is 3000, but from the shell script it doesn't look like its trying to generate more than 3000. 
 
For your reference, this is the script I submitted.  These grid parameters worked fine during correction, but I don't recall a larger array than 1300 jobs.
 

 
#!/bin/bash
#SBATCH --job-name=canu_spiny1-trim
#SBATCH --mail-user=bbarbazuk@ufl.edu
#SBATCH --mem=128gb
#SBATCH -t 96:00:00
#SBATCH --output=outCanu_spiny1-trim%j.out
#SBATCH --error=errCanu_spiny1-trim%j.err
#SBATCH --qos=barbazuk-b 
export PROJ_ROOT=/ufrc/barbazuk/ruthdavenport/ACOMYS-ASSEMBLY/ASSEMBLY_RUNS/Spiny_02132018
 
module load ufrc
module load gcc/5.2.0 
module load gnuplot/4.6.0
module load canu/1.6
module load java/1.8.0_31
 
export _JAVA_OPTIONS="-Xmx16g"
canu -trim \
-p Spiny_02022018 -d Spiny_02022018 \
genomeSize=2.5g \
rawErrorRate=0.300 \
correctedErrorRate=0.035 \
corMhapSensitivity=normal \
corMinCoverage=4 \
corOutCoverage=40 \
minOverlapLength=500 \
minReadLength=1000 \
shell="/bin/bash" \
java="/apps/java/1.8.0_31/bin/java" \
gnuplotTested=true \
gnuplot="/apps/gcc/5.2.0/gnuplot/5.0.3/bin/gnuplot" \
gnuplotImageFormat=svg \
-pacbio-corrected $PROJ_ROOT/Spiny_02132018.correctedReads.fasta.gz \
usegrid=1 \
gridOptionsJobName=CANU-default \
gridOptions="--mem=128gb --time=96:00:00 --qos=barbazuk-b"
 
 
I greatly appreciate any help and suggestions you can provide me.

thanks,


 
 ```

**Comment by skoren** [comment]
I'd guess there is a total job limit of 3000 too which is causing the error since you have more than 3000 jobs total. A couple issues here, first gridOptions is going to always request 128gb of ram even though the overlap jobs are only asking for 4gb so your jobs will be over-requesting memory by 30-fold. The 4gb overlap jobs are pretty small, what does the preamble which detects resources say? I assume there are some small machines which are causing Canu to pick a small memory size to maximize cores used.

You can manually adjust the sizes of the jobs with ```obtOvlHashBlockLength=200000000 utgOvlHashBlockLength= 200000000 obtOvlRefBlockSize=5000000 utgOvlRefBlockSize=5000000  obtOvlMemory=8 utgOvlMemory=8```. Remove the gridOptions memory option (or change gridOptions to gridOptionsExecutive) and the 1-overlapper folder. You could also switch to Canu 1.7 which will use larger jobs for this step by default. You'd have to start the trimming from the corrected reads, you can't reuse any of the current output from 1.6 in 1.7.

**Comment by bbarbazuk** [comment]
Thanks for your response. Yes - there is a limit of 3000, jibs, and this is reiterated in the preamble.  When I look at the Spiny_02022018.partition.err file it seems there are 3221 jobs - so maybe this is the issue.

The corrected reads were produced bu 1.6, but it is OK to use these for trimming under 1.7, or do they need to created in 1.7?

If I can switch to 1,7 then there is no adjust job sizes?  If I continue with 1.6, are the adjustment statements added to my master trimming script?  
Finally, what is your advice regarding memory allocation in the sbatch script that I kick off to run the trimming?  There are a number of hosts with multiple cpus and memory allocations that work out to ~4Gb/core.
Whatever your memory recommendation, would this be specified in the directives of my submission script (#SBATCH --mem=x), or only in the gridOptions passed to canu further down the script, or both?
I can send you my submission script if you think thats helpful.  Thanks for your time!

**Comment by skoren** [comment]
You can use 1.6 corrected reads as input for 1.7 yes.

If you switch to 1.7 then the job sizes should already be larger by default so you wouldn't need to adjust them. If you continue with 1.6 add the parameters I suggested above to your trimming script for Canu. I'd also remove the mem 128gb option from gridOptions to avoid under-scheduling nodes.

Canu handles the memory/threads allocation/request itself so you don't need to adjust it. Your script which kicks off to run trimming will do very little work and submit follow-on jobs so it shouldn't need more than a couple of GB (8gb to be safe).

**Comment by skoren** [comment]
Closing, inactive. 

Another reason to switch to 1.7, Canu 1.6 has a bug when running each of the steps by hand. If you run canu -assemble with a folder already containing trimmed reads and corrected reads, it will assemble the corrected reads not the trimmed reads like it should be. Running each of the three steps in a separate folder would fix this.
